binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+# Binary data formats {#sec-binary-data}
+
+```{r setup}
+#| results: "asis"
+#| echo: false
+source("_common.R")
+status("drafting")
+```
+
+## Introduction
+
+### Prerequisites
+
+This chapter focuses on large data sets, and binary file formats that are well suited to working with them.
+We'll use a data set of item checkouts from Seattle public libraries, available online at [data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6](https://data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6).
+The data is distributed as a single 9GB csv file, so it may take some time to download.
+For the purposes of this chapter we'll assume that you have a local copy of the file at `data/seattle-library-checkout.csv`.
+
+```{r load-packages}
+#| message: false
+#| warning: false
+library(tidyverse)
+library(arrow)
+library(duckdb)
+library(dbplyr)
+```
+
+## The arrow package
+
+In this chapter we'll rely on Apache Arrow ([arrow.apache.org](https://arrow.apache.org/)), a multi-language toolbox designed for efficient analysis and transport of large data sets.
+The arrow package provides a standard way to use Apache Arrow in R, and allows you to analyze larger-than-memory datasets using familiar dplyr syntax.
+As an additional benefit, arrow is extremely fast: you'll see some examples later in the chapter.
+
+### Opening a dataset
+
+Let's start by taking a look at the Seattle library checkout data.
+At 9GB in size, the csv file is large enough that we probably don't want to load the whole thing into memory.
+Instead of using `read_csv()` we'll use the `open_dataset()` function from the arrow package:
+
+The data set is large enough that we might not want to try loading the entire thing into memory.
+Instead, we'll use `arrow::open_dataset()` to scan the file on-disk and create an Arrow Dataset object called `seattle_lib`:
+
+```{r open-dataset}
+seattle_lib <- open_dataset(
+  sources = "data/seattle-library-checkouts.csv", 
+  format = "csv"
+)
+```
+
+When this code is run, `open_dataset()` scans the file to work out its structure and creates the `seattle_lib` object that stores metadata.
+The data remain on-disk, and are not loaded into memory.
+Nevertheless, we are able to call dplyr functions to interact with the data:
+
+```{r glimpse-data}
+#| cache: true
+glimpse(seattle_lib)
```

Comment:
What does it look like if you just print it?

Reply:
The print method prints column names and types (i.e. the schema) but does not print any data values.

Reply:
My qualm with glimpse is that I don't think we use it elsewhere in the book, but maybe I should fix that.

Reply:
I use `glimpse()` pretty regularly in teaching, so I think it's good to add to other places in the book.

Reply:
+1 for using glimpse a little more in the book. I use glimpse a lot in writing posts and (used to, when I taught) find it really helpful for teaching purposes too. Happy to remove it here though if that's preferable

Reply:
I've added it in #1115, I'd say should be good to leave in

Reply:
I also added it to https://r4ds.hadley.nz/data-transform.html#nycflights13